File: impl/ruby-claudecode-sonnet4-1757425857/lib/elf_lang/parser.rb
module ElfLang
  class Parser
    def initialize(tokens)
      @tokens = tokens
      @pos = 0
    end
    
    def parse
      statements = []
      
      while @pos < @tokens.length
        stmt = parse_statement
        statements << stmt if stmt
      end
      
      {
        "type" => "Program",
        "statements" => statements
      }
    end
    
    private
    
    def current_token
      @pos < @tokens.length ? @tokens[@pos] : nil
    end
    
    def peek_token(offset = 1)
      pos = @pos + offset
      pos < @tokens.length ? @tokens[pos] : nil
    end
    
    def advance
      @pos += 1
    end
    
    def consume(expected_type)
      token = current_token
      if token && token['type'] == expected_type
        advance
        token
      else
        raise "Expected #{expected_type}, got #{token ? token['type'] : 'EOF'}"
      end
    end
    
    def parse_statement
      return nil unless current_token
      
      # Handle comments
      if current_token['type'] == 'CMT'
        comment_value = current_token['value']
        advance
        return {
          "type" => "Comment",
          "value" => comment_value
        }
      end
      
      expr = parse_expression
      
      # Optional semicolon
      if current_token && current_token['type'] == ';'
        advance
      end
      
      {
        "type" => "Expression",
        "value" => expr
      }
    end
    
    def parse_expression
      parse_let_or_expression
    end
    
    def parse_let_or_expression
      if current_token && current_token['type'] == 'LET'
        advance
        
        # Check for 'mut' keyword
        is_mutable = false
        if current_token && current_token['type'] == 'MUT'
          is_mutable = true
          advance
        end
        
        # Get identifier name
        name_token = consume('ID')
        name = {
          "type" => "Identifier",
          "name" => name_token['value']
        }
        
        # Consume '='
        consume('=')
        
        # Parse the value expression
        value = parse_assignment_expression
        
        {
          "type" => is_mutable ? "MutableLet" : "Let",
          "name" => name,
          "value" => value
        }
      else
        parse_assignment_expression
      end
    end
    
    def parse_assignment_expression
      expr = parse_or_expression
      
      # Check for assignment: identifier = value
      if current_token && current_token['type'] == '=' && expr['type'] == 'Identifier'
        advance  # consume '='
        value = parse_assignment_expression  # Right-associative
        
        return {
          "type" => "Assignment",
          "name" => expr,
          "value" => value
        }
      end
      
      expr
    end
    
    def parse_or_expression
      left = parse_and_expression
      
      while current_token && current_token['type'] == '||'
        op = current_token['value']
        advance
        right = parse_and_expression
        left = {
          "type" => "Infix",
          "operator" => op,
          "left" => left,
          "right" => right
        }
      end
      
      left
    end
    
    def parse_and_expression
      left = parse_equality_expression
      
      while current_token && current_token['type'] == '&&'
        op = current_token['value']
        advance
        right = parse_equality_expression
        left = {
          "type" => "Infix",
          "operator" => op,
          "left" => left,
          "right" => right
        }
      end
      
      left
    end
    
    def parse_equality_expression
      left = parse_comparison_expression
      
      while current_token && ['==', '!='].include?(current_token['type'])
        op = current_token['value']
        advance
        right = parse_comparison_expression
        left = {
          "type" => "Infix",
          "operator" => op,
          "left" => left,
          "right" => right
        }
      end
      
      left
    end
    
    def parse_comparison_expression
      left = parse_thread_expression
      
      while current_token && ['>', '<', '>=', '<='].include?(current_token['type'])
        op = current_token['value']
        advance
        right = parse_thread_expression
        left = {
          "type" => "Infix",
          "operator" => op,
          "left" => left,
          "right" => right
        }
      end
      
      left
    end
    
    def parse_thread_expression
      left = parse_compose_expression
      
      if current_token && current_token['type'] == '|>'
        # Collect all threaded functions
        functions = []
        
        while current_token && current_token['type'] == '|>'
          advance
          functions << parse_compose_expression
        end
        
        {
          "type" => "FunctionThread",
          "initial" => left,
          "functions" => functions
        }
      else
        left
      end
    end
    
    def parse_compose_expression
      left = parse_additive_expression
      
      if current_token && current_token['type'] == '>>'
        # Collect all composed functions
        functions = [left]
        
        while current_token && current_token['type'] == '>>'
          advance
          functions << parse_additive_expression
        end
        
        {
          "type" => "FunctionComposition",
          "functions" => functions
        }
      else
        left
      end
    end
    
    def parse_additive_expression
      left = parse_multiplicative_expression
      
      while current_token && ['+', '-'].include?(current_token['type'])
        op = current_token['value']
        advance
        right = parse_multiplicative_expression
        left = {
          "type" => "Infix",
          "operator" => op,
          "left" => left,
          "right" => right
        }
      end
      
      left
    end
    
    def parse_multiplicative_expression
      left = parse_unary_expression
      
      while current_token && ['*', '/'].include?(current_token['type'])
        op = current_token['value']
        advance
        right = parse_unary_expression
        left = {
          "type" => "Infix",
          "operator" => op,
          "left" => left,
          "right" => right
        }
      end
      
      left
    end
    
    def parse_unary_expression
      if current_token && current_token['type'] == '-'
        advance
        expr = parse_unary_expression
        {
          "type" => "Unary",
          "operator" => "-",
          "operand" => expr
        }
      else
        parse_postfix_expression
      end
    end
    
    def parse_postfix_expression
      left = parse_primary_expression
      
      while true
        case current_token&.[]('type')
        when '['
          # Indexing
          advance
          index = parse_expression
          consume(']')
          left = {
            "type" => "Index",
            "left" => left,
            "index" => index
          }
        when '('
          # Function call
          advance
          args = []
          
          unless current_token && current_token['type'] == ')'
            loop do
              args << parse_expression
              
              if current_token && current_token['type'] == ','
                advance
              else
                break
              end
            end
          end
          
          consume(')')
          left = {
            "type" => "Call",
            "function" => left,
            "arguments" => args
          }
        else
          break
        end
      end
      
      left
    end
    
    def parse_primary_expression
      return nil unless current_token
      
      case current_token['type']
      when 'INT'
        value = current_token['value']
        advance
        {
          "type" => "Integer",
          "value" => value
        }
      when 'DEC'
        value = current_token['value']
        advance
        {
          "type" => "Decimal",
          "value" => value
        }
      when 'STR'
        raw_value = current_token['value']
        # Remove surrounding quotes and handle escapes
        content = raw_value[1...-1]  # Remove quotes
        content = content.gsub('\\n', "\n").gsub('\\t', "\t").gsub('\\"', '"').gsub('\\\\', '\\')
        advance
        {
          "type" => "String",
          "value" => content
        }
      when 'TRUE'
        advance
        {
          "type" => "Boolean",
          "value" => true
        }
      when 'FALSE'
        advance
        {
          "type" => "Boolean",
          "value" => false
        }
      when 'NIL'
        advance
        {
          "type" => "Nil"
        }
      when 'ID'
        name = current_token['value']
        advance
        {
          "type" => "Identifier",
          "name" => name
        }
      when '+', '-', '*', '/'
        name = current_token['value']
        advance
        {
          "type" => "Identifier",
          "name" => name
        }
      when '('
        advance
        expr = parse_expression
        consume(')')
        expr
      when '['
        parse_list_literal
      when '{'
        parse_set_literal
      when '#{'
        parse_dict_literal
      when '|'
        parse_function_literal
      when '||'
        parse_empty_function_literal
      when 'IF'
        parse_if_expression
      else
        raise "Unexpected token: #{current_token['type']}"
      end
    end
    
    def parse_list_literal
      advance  # consume '['
      elements = []
      
      unless current_token && current_token['type'] == ']'
        loop do
          elements << parse_expression
          
          if current_token && current_token['type'] == ','
            advance
          else
            break
          end
        end
      end
      
      consume(']')
      
      {
        "type" => "List",
        "items" => elements
      }
    end
    
    def parse_set_literal
      advance  # consume '{'
      elements = []
      
      unless current_token && current_token['type'] == '}'
        loop do
          elements << parse_expression
          
          if current_token && current_token['type'] == ','
            advance
          else
            break
          end
        end
      end
      
      consume('}')
      
      {
        "type" => "Set",
        "items" => elements
      }
    end
    
    def parse_dict_literal
      advance  # consume '#{'
      pairs = []
      
      unless current_token && current_token['type'] == '}'
        loop do
          key = parse_expression
          consume(':')
          value = parse_expression
          
          pairs << {
            "key" => key,
            "value" => value
          }
          
          if current_token && current_token['type'] == ','
            advance
          else
            break
          end
        end
      end
      
      consume('}')
      
      {
        "type" => "Dictionary",
        "items" => pairs
      }
    end
    
    def parse_function_literal
      advance  # consume '|'
      parameters = []
      
      unless current_token && current_token['type'] == '|'
        loop do
          param_token = consume('ID')
          parameters << {
            "type" => "Identifier",
            "name" => param_token['value']
          }
          
          if current_token && current_token['type'] == ','
            advance
          else
            break
          end
        end
      end
      
      consume('|')
      
      # Check if we have a block or a simple expression
      if current_token && current_token['type'] == '{'
        body = parse_block
      else
        # Wrap expression in a Block
        expr = parse_expression
        body = {
          "type" => "Block",
          "statements" => [
            {
              "type" => "Expression",
              "value" => expr
            }
          ]
        }
      end
      
      {
        "type" => "Function",
        "parameters" => parameters,
        "body" => body
      }
    end
    
    def parse_empty_function_literal
      advance  # consume '||'
      parameters = []  # No parameters for ||
      
      # Check if we have a block or a simple expression
      if current_token && current_token['type'] == '{'
        body = parse_block
      else
        # Wrap expression in a Block
        expr = parse_expression
        body = {
          "type" => "Block",
          "statements" => [
            {
              "type" => "Expression",
              "value" => expr
            }
          ]
        }
      end
      
      {
        "type" => "Function",
        "parameters" => parameters,
        "body" => body
      }
    end
    
    def parse_if_expression
      advance  # consume 'IF'
      
      condition = parse_expression
      then_branch = parse_block
      
      else_branch = nil
      if current_token && current_token['type'] == 'ELSE'
        advance
        else_branch = parse_block
      end
      
      {
        "type" => "If",
        "condition" => condition,
        "consequence" => then_branch,
        "alternative" => else_branch
      }
    end
    
    def parse_block
      consume('{')
      statements = []
      
      while current_token && current_token['type'] != '}'
        stmt = parse_statement
        statements << stmt if stmt
      end
      
      consume('}')
      
      {
        "type" => "Block",
        "statements" => statements
      }
    end
  end
end

File: impl/fsharp-claudecode-sonnet4-1757761598/Parser.fs
module ElfLang.Parser

open ElfLang.Lexer
open System.Text.Json
open System.Collections.Generic

// AST Types based on test expectations
type Expression =
    | Integer of value: string
    | Decimal of value: string
    | String of value: string
    | Boolean of value: bool
    | Nil
    | Identifier of name: string * ``type``: string
    | Let of name: Expression * value: Expression
    | MutableLet of name: Expression * value: Expression
    | Assignment of name: Expression * value: Expression
    | Infix of left: Expression * operator: string * right: Expression
    | Unary of operator: string * operand: Expression
    | Index of target: Expression * index: Expression
    | List of elements: Expression list
    | Set of elements: Expression list
    | Dictionary of entries: (Expression * Expression) list
    | If of condition: Expression * thenBranch: Block * elseBranch: Block option
    | Function of parameters: Expression list * body: Block
    | Call of func: Expression * arguments: Expression list
    | FunctionThread of initial: Expression * functions: Expression list
    | FunctionComposition of functions: Expression list

and Statement =
    | Expression of value: Expression
    | Comment of value: string

and Block = {
    statements: Statement list
    blockType: string  // Always "Block"
}

and Program = {
    statements: Statement list
    ``type``: string  // Always "Program"
}

// Token parser state
type Parser = {
    tokens: Token array
    current: int
}

// Parser helper functions
let peek (parser: Parser) =
    if parser.current < parser.tokens.Length then
        Some parser.tokens[parser.current]
    else
        None

let advance (parser: Parser) =
    { parser with current = parser.current + 1 }

let consume (parser: Parser) (expectedType: TokenType) =
    match peek parser with
    | Some token when token.Type = expectedType ->
        (token, advance parser)
    | Some token ->
        failwith $"Expected {expectedType}, found {token.Type}"
    | None ->
        failwith $"Expected {expectedType}, found EOF"

let check (parser: Parser) (tokenType: TokenType) =
    match peek parser with
    | Some token -> token.Type = tokenType
    | None -> false

let isAtEnd (parser: Parser) =
    parser.current >= parser.tokens.Length

// Precedence levels for operators (higher number = higher precedence)
let getOperatorPrecedence (op: string) =
    match op with
    | "|>" -> 0    // Threading has lowest precedence
    | "||" -> 1
    | "&&" -> 2
    | ">>" -> 2    // Function composition
    | "==" | "!=" | ">" | "<" | ">=" | "<=" -> 3
    | "+" | "-" -> 4
    | "*" | "/" -> 5
    | _ -> 0

// Parse primary expressions (literals, identifiers, parentheses)
let rec parsePrimary (parser: Parser) : Expression * Parser =
    match peek parser with
    | Some { Type = INT; Value = value } ->
        (Integer value, advance parser)
    | Some { Type = DEC; Value = value } ->
        (Decimal value, advance parser)
    | Some { Type = STR; Value = value } ->
        // Remove quotes and unescape string value for AST
        let quotesRemoved = value.Substring(1, value.Length - 2)
        let unescapedValue = quotesRemoved.Replace("\\\"", "\"").Replace("\\\\", "\\").Replace("\\n", "\n").Replace("\\t", "\t")
        (String unescapedValue, advance parser)
    | Some { Type = TRUE; Value = _ } ->
        (Boolean true, advance parser)
    | Some { Type = FALSE; Value = _ } ->
        (Boolean false, advance parser)
    | Some { Type = NIL; Value = _ } ->
        (Nil, advance parser)
    | Some { Type = ID; Value = name } ->
        (Identifier (name, "Identifier"), advance parser)
    | Some token when (token.Type = PLUS || token.Type = MINUS) ->
        // Check if this is a unary operator or an operator used as identifier
        let parser = advance parser
        // If followed by a primary expression (but NOT LPAREN for function calls), treat as unary operator
        if not (isAtEnd parser) then
            let nextToken = peek parser
            match nextToken with
            | Some { Type = LPAREN } ->
                // Function call like +(1, 2), treat as identifier
                (Identifier (token.Value, "Identifier"), parser)
            | Some { Type = INT | DEC | STR | TRUE | FALSE | NIL | ID | LBRACKET | LBRACE | LHASHBRACE | PIPE | IF | LET } ->
                // This is a unary operator
                let (operand, parser) = parsePrimary parser
                (Unary (token.Value, operand), parser)
            | _ ->
                // Not followed by a primary expression, treat as identifier
                (Identifier (token.Value, "Identifier"), parser)
        else
            // End of input, treat as identifier
            (Identifier (token.Value, "Identifier"), parser)
    | Some token when (token.Type = MULT || token.Type = DIV) ->
        // Multiplication and division are always identifiers (can't be unary)
        (Identifier (token.Value, "Identifier"), advance parser)
    | Some { Type = LPAREN; Value = _ } ->
        let parser = advance parser
        let (expr, parser) = parseExpression parser
        let (_, parser) = consume parser RPAREN
        (expr, parser)
    | Some { Type = LBRACKET; Value = _ } ->
        parseList parser
    | Some { Type = LBRACE; Value = _ } ->
        parseSet parser
    | Some { Type = LHASHBRACE; Value = _ } ->
        parseDictionary parser
    | Some { Type = PIPE; Value = _ } ->
        parseFunction parser
    | Some { Type = OR_OR; Value = _ } ->
        parseEmptyParameterFunction parser
    | Some { Type = IF; Value = _ } ->
        parseIf parser
    | Some { Type = LET; Value = _ } ->
        parseLet parser
    | Some token ->
        failwith $"Unexpected token in primary expression: {token.Type} '{token.Value}'"
    | None ->
        failwith "Expected expression, found EOF"

// Parse list literals [1, 2, 3]
and parseList (parser: Parser) : Expression * Parser =
    let (_, parser) = consume parser LBRACKET
    let rec parseElements parser acc =
        if check parser RBRACKET then
            (List.rev acc, parser)
        else
            let (expr, parser) = parseExpression parser
            let acc = expr :: acc
            if check parser COMMA then
                let parser = advance parser
                parseElements parser acc
            else
                (List.rev acc, parser)

    let (elements, parser) = parseElements parser []
    let (_, parser) = consume parser RBRACKET
    (List elements, parser)

// Parse set literals {1, 2, 3}
and parseSet (parser: Parser) : Expression * Parser =
    let (_, parser) = consume parser LBRACE
    let rec parseElements parser acc =
        if check parser RBRACE then
            (List.rev acc, parser)
        else
            let (expr, parser) = parseExpression parser
            let acc = expr :: acc
            if check parser COMMA then
                let parser = advance parser
                parseElements parser acc
            else
                (List.rev acc, parser)

    let (elements, parser) = parseElements parser []
    let (_, parser) = consume parser RBRACE
    (Set elements, parser)

// Parse dictionary literals #{"a": 1, "b": 2}
and parseDictionary (parser: Parser) : Expression * Parser =
    let (_, parser) = consume parser LHASHBRACE
    let rec parseEntries parser acc =
        if check parser RBRACE then
            (List.rev acc, parser)
        else
            let (key, parser) = parseExpression parser
            let (_, parser) = consume parser COLON // Need to add COLON token
            let (value, parser) = parseExpression parser
            let acc = (key, value) :: acc
            if check parser COMMA then
                let parser = advance parser
                parseEntries parser acc
            else
                (List.rev acc, parser)

    let (entries, parser) = parseEntries parser []
    let (_, parser) = consume parser RBRACE
    (Dictionary entries, parser)

// Parse function literals |x, y| expr or |x| { ... }
and parseFunction (parser: Parser) : Expression * Parser =
    let (_, parser) = consume parser PIPE

    // Parse parameters
    let rec parseParameters parser acc =
        if check parser PIPE then
            (List.rev acc, parser)
        else
            match peek parser with
            | Some { Type = ID; Value = name } ->
                let parser = advance parser
                let param = Identifier (name, "Identifier")
                let acc = param :: acc
                if check parser COMMA then
                    let parser = advance parser
                    parseParameters parser acc
                else
                    (List.rev acc, parser)
            | _ ->
                failwith "Expected parameter name"

    let (parameters, parser) = parseParameters parser []
    let (_, parser) = consume parser PIPE

    // Parse body - could be expression or block
    let (body, parser) =
        if check parser LBRACE then
            parseBlock parser
        else
            // Single expression - wrap in block
            let (expr, parser) = parseExpression parser
            let stmt = Expression expr
            let block = { statements = [stmt]; blockType = "Block" }
            (block, parser)

    (Function (parameters, body), parser)

// Parse empty parameter function || { ... }
and parseEmptyParameterFunction (parser: Parser) : Expression * Parser =
    let (_, parser) = consume parser OR_OR

    // Parse body - should be a block
    let (body, parser) =
        if check parser LBRACE then
            parseBlock parser
        else
            // Single expression - wrap in block
            let (expr, parser) = parseExpression parser
            let stmt = Expression expr
            let block = { statements = [stmt]; blockType = "Block" }
            (block, parser)

    (Function ([], body), parser)

// Parse if expressions
and parseIf (parser: Parser) : Expression * Parser =
    let (_, parser) = consume parser IF
    let (condition, parser) = parseExpression parser
    let (thenBlock, parser) = parseBlock parser
    let (elseBlock, parser) =
        if check parser ELSE then
            let parser = advance parser
            let (block, parser) = parseBlock parser
            (Some block, parser)
        else
            (None, parser)

    (If (condition, thenBlock, elseBlock), parser)

// Parse let declarations
and parseLet (parser: Parser) : Expression * Parser =
    let (_, parser) = consume parser LET
    let isMutable = check parser MUT
    let parser = if isMutable then advance parser else parser

    match peek parser with
    | Some { Type = ID; Value = name } ->
        let parser = advance parser
        let (_, parser) = consume parser ASSIGN
        let (value, parser) = parseExpression parser
        let identifier = Identifier (name, "Identifier")
        let expr = if isMutable then MutableLet (identifier, value) else Let (identifier, value)
        (expr, parser)
    | _ ->
        failwith "Expected identifier after 'let'"

// Parse block { ... }
and parseBlock (parser: Parser) : Block * Parser =
    let (_, parser) = consume parser LBRACE
    let rec parseStatements parser acc =
        if check parser RBRACE then
            (List.rev acc, parser)
        else
            let (stmt, parser) = parseStatement parser
            let acc = stmt :: acc
            // Optional semicolon
            let parser = if check parser SEMICOLON then advance parser else parser
            parseStatements parser acc

    let (statements, parser) = parseStatements parser []
    let (_, parser) = consume parser RBRACE
    ({ statements = statements; blockType = "Block" }, parser)

// Parse binary expressions with precedence
and parseBinaryExpression (parser: Parser) (left: Expression) (minPrecedence: int) : Expression * Parser =
    let rec loop parser left =
        match peek parser with
        | Some token when (token.Type = PLUS || token.Type = MINUS || token.Type = MULT || token.Type = DIV ||
                          token.Type = EQ_EQ || token.Type = NOT_EQ || token.Type = GT || token.Type = LT ||
                          token.Type = GTE || token.Type = LTE || token.Type = AND_AND || token.Type = OR_OR ||
                          token.Type = PIPE_GT || token.Type = GT_GT) ->
            let operator = token.Value
            let precedence = getOperatorPrecedence operator
            if precedence >= minPrecedence then
                let parser = advance parser
                let (right, parser) = parsePrimary parser
                let (right, parser) = parsePostfix parser right
                // Use precedence + 1 for left-associative operators
                let (right, parser) = parseBinaryExpression parser right (precedence + 1)

                // Create appropriate AST node based on operator
                let expr =
                    match operator with
                    | "|>" ->
                        // Handle threading - collect all functions in the chain
                        match left with
                        | FunctionThread (initial, funcs) -> FunctionThread (initial, funcs @ [right])
                        | _ -> FunctionThread (left, [right])
                    | ">>" ->
                        // Handle function composition - collect all functions in the chain
                        match left with
                        | FunctionComposition funcs -> FunctionComposition (funcs @ [right])
                        | _ -> FunctionComposition [left; right]
                    | _ -> Infix (left, operator, right)

                loop parser expr
            else
                (left, parser)
        | _ ->
            (left, parser)

    loop parser left

// Parse postfix expressions (like indexing and function calls)
and parsePostfix (parser: Parser) (expr: Expression) : Expression * Parser =
    match peek parser with
    | Some { Type = LBRACKET; Value = _ } ->
        let parser = advance parser
        let (index, parser) = parseExpression parser
        let (_, parser) = consume parser RBRACKET
        let indexExpr = Index (expr, index)
        parsePostfix parser indexExpr
    | Some { Type = LPAREN; Value = _ } ->
        // Function call: expr(arg1, arg2, ...)
        let parser = advance parser
        let rec parseArguments parser acc =
            if check parser RPAREN then
                (List.rev acc, parser)
            else
                let (arg, parser) = parseExpression parser
                let acc = arg :: acc
                if check parser COMMA then
                    let parser = advance parser
                    parseArguments parser acc
                else
                    (List.rev acc, parser)

        let (arguments, parser) = parseArguments parser []
        let (_, parser) = consume parser RPAREN
        let callExpr = Call (expr, arguments)
        parsePostfix parser callExpr
    | _ ->
        (expr, parser)

// Parse expressions
and parseExpression (parser: Parser) : Expression * Parser =
    let (left, parser) = parsePrimary parser
    let (left, parser) = parsePostfix parser left
    parseBinaryExpression parser left 0

// Parse statements
and parseStatement (parser: Parser) : Statement * Parser =
    // Check for assignment: identifier = expression
    if check parser ID then
        let identifierToken = parser.tokens[parser.current]
        let parser = advance parser
        if check parser ASSIGN then
            let parser = advance parser
            let (value, parser) = parseExpression parser
            let identifier = Identifier (identifierToken.Value, "Identifier")
            let assignmentExpr = Assignment (identifier, value)
            (Expression assignmentExpr, parser)
        else
            // Not an assignment, backtrack and parse as normal expression
            let parser = { parser with current = parser.current - 1 }
            let (expr, parser) = parseExpression parser
            (Expression expr, parser)
    else
        let (expr, parser) = parseExpression parser
        (Expression expr, parser)

// Main parse function
let parse (tokens: Token list) : Program =
    let parser = { tokens = List.toArray tokens; current = 0 }

    let rec parseStatements parser acc =
        if isAtEnd parser then
            List.rev acc
        else
            match peek parser with
            | Some { Type = CMT; Value = value } ->
                // Parse comment as a statement
                let parser = advance parser
                let commentStmt = Comment value
                let acc = commentStmt :: acc
                parseStatements parser acc
            | _ ->
                let (stmt, parser) = parseStatement parser
                let acc = stmt :: acc
                // Optional semicolon
                let parser = if check parser SEMICOLON then advance parser else parser
                parseStatements parser acc

    let statements = parseStatements parser []
    { statements = statements; ``type`` = "Program" }

// JSON Serialization functions for AST (manual approach for proper formatting)
let rec expressionToJson (expr: Expression) : string =
    match expr with
    | Integer value ->
        $"{{\"type\":\"Integer\",\"value\":\"{value}\"}}"
    | Decimal value ->
        $"{{\"type\":\"Decimal\",\"value\":\"{value}\"}}"
    | String value ->
        // The value is already unescaped from the lexer, just need to escape for JSON
        let escapedValue = value.Replace("\\", "\\\\").Replace("\"", "\\\"")
        $"{{\"type\":\"String\",\"value\":\"{escapedValue}\"}}"
    | Boolean value ->
        sprintf "{\"type\":\"Boolean\",\"value\":%s}" (if value then "true" else "false")
    | Nil ->
        "{\"type\":\"Nil\"}"
    | Identifier (name, _) ->
        $"{{\"name\":\"{name}\",\"type\":\"Identifier\"}}"
    | Let (name, value) ->
        let nameJson = expressionToJson name
        let valueJson = expressionToJson value
        $"{{\"name\":{nameJson},\"type\":\"Let\",\"value\":{valueJson}}}"
    | MutableLet (name, value) ->
        let nameJson = expressionToJson name
        let valueJson = expressionToJson value
        $"{{\"name\":{nameJson},\"type\":\"MutableLet\",\"value\":{valueJson}}}"
    | Assignment (name, value) ->
        let nameJson = expressionToJson name
        let valueJson = expressionToJson value
        $"{{\"name\":{nameJson},\"type\":\"Assignment\",\"value\":{valueJson}}}"
    | Infix (left, op, right) ->
        let leftJson = expressionToJson left
        let rightJson = expressionToJson right
        $"{{\"left\":{leftJson},\"operator\":\"{op}\",\"right\":{rightJson},\"type\":\"Infix\"}}"
    | Unary (op, operand) ->
        let operandJson = expressionToJson operand
        $"{{\"operand\":{operandJson},\"operator\":\"{op}\",\"type\":\"Unary\"}}"
    | List elements ->
        let elementsJson = elements |> List.map expressionToJson |> String.concat ","
        $"{{\"items\":[{elementsJson}],\"type\":\"List\"}}"
    | Set elements ->
        let elementsJson = elements |> List.map expressionToJson |> String.concat ","
        $"{{\"items\":[{elementsJson}],\"type\":\"Set\"}}"
    | Dictionary entries ->
        let entriesJson = entries |> List.map (fun (k, v) -> $"{{\"key\":{expressionToJson k},\"value\":{expressionToJson v}}}") |> String.concat ","
        $"{{\"items\":[{entriesJson}],\"type\":\"Dictionary\"}}"
    | Function (parameters, body) ->
        let parametersJson = parameters |> List.map expressionToJson |> String.concat ","
        let bodyJson = blockToJson body
        $"{{\"body\":{bodyJson},\"parameters\":[{parametersJson}],\"type\":\"Function\"}}"
    | If (condition, thenBranch, elseBranch) ->
        let conditionJson = expressionToJson condition
        let thenJson = blockToJson thenBranch
        match elseBranch with
        | Some elseBlock ->
            let elseJson = blockToJson elseBlock
            $"{{\"alternative\":{elseJson},\"condition\":{conditionJson},\"consequence\":{thenJson},\"type\":\"If\"}}"
        | None ->
            $"{{\"condition\":{conditionJson},\"consequence\":{thenJson},\"type\":\"If\"}}"
    | Call (func, arguments) ->
        let funcJson = expressionToJson func
        let argsJson = arguments |> List.map expressionToJson |> String.concat ","
        $"{{\"arguments\":[{argsJson}],\"function\":{funcJson},\"type\":\"Call\"}}"
    | Index (target, index) ->
        let targetJson = expressionToJson target
        let indexJson = expressionToJson index
        $"{{\"index\":{indexJson},\"left\":{targetJson},\"type\":\"Index\"}}"
    | FunctionThread (initial, functions) ->
        let initialJson = expressionToJson initial
        let functionsJson = functions |> List.map expressionToJson |> String.concat ","
        $"{{\"functions\":[{functionsJson}],\"initial\":{initialJson},\"type\":\"FunctionThread\"}}"
    | FunctionComposition (functions) ->
        let functionsJson = functions |> List.map expressionToJson |> String.concat ","
        $"{{\"functions\":[{functionsJson}],\"type\":\"FunctionComposition\"}}"

and blockToJson (block: Block) : string =
    let statementsJson = block.statements |> List.map statementToJson |> String.concat ","
    $"{{\"statements\":[{statementsJson}],\"type\":\"Block\"}}"

and statementToJson (stmt: Statement) : string =
    match stmt with
    | Expression expr ->
        let exprJson = expressionToJson expr
        $"{{\"type\":\"Expression\",\"value\":{exprJson}}}"
    | Comment value ->
        $"{{\"type\":\"Comment\",\"value\":\"{value}\"}}"

let programToJson (program: Program) : string =
    let statementsJson = program.statements |> List.map statementToJson |> String.concat ","
    $"{{\"statements\":[{statementsJson}],\"type\":\"Program\"}}"

// Pretty print JSON with 2-space indentation
let prettyPrintJson (json: string) : string =
    let doc = JsonDocument.Parse(json)
    use stream = new System.IO.MemoryStream()
    let options = JsonWriterOptions(Indented = true, Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping)
    use writer = new Utf8JsonWriter(stream, options)
    doc.WriteTo(writer)
    writer.Flush()
    let result = System.Text.Encoding.UTF8.GetString(stream.ToArray())

    // Convert from 4-space to 2-space indentation - handle all levels properly
    let lines = result.Split('\n')
    let convertedLines = lines |> Array.map (fun line ->
        let leadingSpaces = line.Length - line.TrimStart(' ').Length
        let newLeadingSpaces = leadingSpaces / 2
        String.replicate newLeadingSpaces "  " + line.TrimStart(' ')
    )
    let converted = String.concat "\n" convertedLines

    // Fix Unicode surrogate pairs back to proper Unicode characters
    let unicodeFixed = System.Text.RegularExpressions.Regex.Replace(
        converted,
        @"\\u([0-9A-Fa-f]{4})\\u([0-9A-Fa-f]{4})",
        fun m ->
            let high = System.Convert.ToInt32(m.Groups.[1].Value, 16)
            let low = System.Convert.ToInt32(m.Groups.[2].Value, 16)
            if high >= 0xD800 && high <= 0xDBFF && low >= 0xDC00 && low <= 0xDFFF then
                // This is a surrogate pair, convert to actual Unicode character
                let codepoint = 0x10000 + ((high &&& 0x3FF) <<< 10) + (low &&& 0x3FF)
                System.Char.ConvertFromUtf32(codepoint)
            else
                m.Value
    )
    unicodeFixed